WalletWasabi.Tests/UnitTests/Blockchain/TransactionOutputs/CoinsRegistryTests.cs

```diff
@@ -0,0 +1,182 @@
+using NBitcoin;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using WalletWasabi.Blockchain.Keys;
+using WalletWasabi.Blockchain.TransactionOutputs;
+using WalletWasabi.Blockchain.Transactions;
+using WalletWasabi.Models;
+using Xunit;
+
+namespace WalletWasabi.Tests.UnitTests.Blockchain.TransactionOutputs;
+
+/// <summary>
+/// Tests for <see cref="CoinsRegistry"/> class.
+/// </summary>
+public class CoinsRegistryTests
+{
+	public CoinsRegistryTests()
+	{
+		KeyManager = KeyManager.CreateNew(out _, password: "password", Network.Main);
+	}
+
+	private KeyManager KeyManager { get; }
+	private CoinsRegistry Coins { get; } = new();
+
+	/// <summary>
+	/// Tests <see cref="CoinsRegistry.Undo(uint256)"/> method.
+	/// </summary>
+	[Fact]
+	public void UndoTransaction()
+	{
+		using CancellationTokenSource testDeadlineCts = new(TimeSpan.FromSeconds(30));
+
+		// --tx0---> (A) --tx1 (replaceable)-+--> (B) --tx2---> (D)
+		//                  |                |
+		//                  |                +--> (C)
+		//                  |
+		//                  +--tx3 (replacement)---> (E)
+
+		Coin? tx0Coin;
+		SmartTransaction tx0;
+		SmartTransaction tx1;
+		SmartTransaction tx2;
+		SmartTransaction tx3;
+
+		// Create and process transaction tx0.
+		{
+			tx0 = CreateCreditingTransaction(NewInternalKey(label: "A").P2wpkhScript, Money.Coins(1.0m), height: 54321);
+			tx0Coin = tx0.Transaction.Outputs.AsCoins().First();
+
+			IReadOnlyList<SmartCoin> tx0Coins = ProcessTransaction(tx0);
+			_ = Assert.Single(tx0Coins);
+			_ = Assert.Single(Coins);
```

Comment:
Nit: I don't think we need to discard these.

```suggestion
			Assert.Single(tx0Coins);
			Assert.Single(Coins);
```